modules/webbrowser/src/processors/webbrowserbase.cpp

```diff
@@ -0,0 +1,150 @@
+/*********************************************************************************
+ *
+ * Inviwo - Interactive Visualization Workshop
+ *
+ * Copyright (c) 2024 Inviwo Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <modules/webbrowser/processors/webbrowserbase.h>
+
+#include <inviwo/core/common/inviwoapplication.h>
+
+#include <inviwo/core/processors/processor.h>
+#include <modules/webbrowser/webbrowserclient.h>
+#include <modules/webbrowser/webbrowsermodule.h>
+#include <modules/webbrowser/webbrowserutil.h>
+
+#include <fmt/core.h>
+
+namespace inviwo {
+
+WebBrowserBase::WebBrowserBase(InviwoApplication* app, Processor* processor)
+    : parentProcessor_{processor}
+    , picking_{processor, 1,
+               [this](PickingEvent* p) { cefInteractionHandler_.handlePickingEvent(p); }}
+    , cefToInviwoImageConverter_{picking_.getColor()}
+    , renderHandler_{static_cast<RenderHandlerGL*>(
+          app->getModuleByType<WebBrowserModule>()->getBrowserClient()->GetRenderHandler().get())} {
+
+    // Setup CEF browser
+    auto [windowInfo, browserSettings] = cefutil::getDefaultBrowserSettings();
+    auto browserClient = app->getModuleByType<WebBrowserModule>()->getBrowserClient();
+    // Note that browserClient_ outlives this class so make sure to remove
+    // renderHandler_ in
+    // destructor
+    browser_ = CefBrowserHost::CreateBrowserSync(
+        windowInfo, browserClient, std::string{getSource()}, browserSettings, nullptr, nullptr);
```

Comment:
getSource will always be empty in the constructor? Pass in source as an argument?